### PR TITLE
Add support for tls mode, and --insecure flag \n c…

### PR DESCRIPTION
**Pull request by TotalKrill**
adds tls support via rustls
adds an --insecure flag 
change clap usage to derive instead instead for simpler code

### CODE REVIEW

**Comment by EdJoPaTo** [comment]
I like seeing your efforts there!

Personally I skipped on using the clap derive stuff because of missing autocompletes on all the possible options that can be specified. Not sure what I want to think about that currently.
Newer clap versions use types (`get_one::<Type>`) like the derive stuff also does and that can get out of sync so having them as enums / structs is way more error proof.
I should play around with them again.

Can we split up this PR and do the TLS stuff separately from the clap changes? (Or one after the other?) That results in less complex diffs to understand and think about.

And thanks again for taking your time providing some improvements, I like to see that and want to incorporate them!

**Comment by TotalKrill** [comment]
Thx for the feedback, I splitted ( splat? ) the commit into two, for a bit easier review, I noticed that one extra argument slank in when moving to clap, but I hope that is fine ( client-id setting )

**Review by EdJoPaTo — CHANGES_REQUESTED** [review]
Sorry for the late reply. Time currently doesnt seem to be on my side. :disappointed: 

I [played around](https://github.com/EdJoPaTo/project-below/commit/3f0c3a32d49fb412736476b2659c037d11dda26e) a little with `clap_derive` and found some issues with the refactoring part of this PR. Personally I think this PR does too much at the same time.

Thinking about the clap derive vs arguments I think its nicer to use the struct in code but its harder to create the `#[clap(…)]` stuff as there is no language server support like autocompletion or documentation for that kind of stuff.
But `cargo doc --open -p clap` can be used for that purpose too its just not as handy. This is its own discussion independent from the main thing this PR wants to introduce: tls mode.
We should get these discussions separated into multiple PRs.

On the other hand the CLI should change towards something supporting WebSockets for example (see #43) and we need to make sure that all that stuff works together somehow. Looking [into the code there](https://github.com/bytebeamio/rumqtt/blob/96f12c7c0a8c73ddaf0635f977100b56d576811e/rumqttc/src/lib.rs#L608-L610) it seems like encryption is always disabled then and needs to be handled then again? I think we should improve the logic of using different ways of connecting to a broker in rumqttc first so all projects depending on can use these improvements.

Do you have any thoughts on that? Sorry for making your PR so complicated… :disappointed:

**Comment by TotalKrill** [comment]
I read your comment as that there is still some undecided things regarding the derive macro switch, I find it much more cleaner to use, and less error-prone to get the desired functionality. Reading your change requests, I agree to all points. 

Lets keep this PR open, I took the tls addition and added to another PR #67 , maybe I can come back clap derive refactor at a later date.

**Comment by EdJoPaTo** [comment]
Awesome, thank you! I will try to find some time to look over it.

File: src/cli.rs [hunk 1]
```diff
@@ -1,161 +1,138 @@
...
+    #[clap(visible_alias = "l")]
+    Log {
+        /// Topics to log
+        #[clap(
+            required = true,
+            env = "MQTT_TOPIC",
+            name = "TOPIC",
+            default_value = "#"
+        )]
+        topics: Vec<String>,
```
[the remark this answers is not in the document]
Comment (TotalKrill) [reply src/cli.rs:37]: This is a good example of why i much prefer the derive macro, changing this name to topic from topics, breaks code silently in the arg way, but now, I got a compilation error in main.rs. I enjoy those nice QoL things a lot

**Comment by TotalKrill** [comment]
I have now remade the both features added in this PR into two separate ones (#69, #67). closing this

**Pull request closed without merging**